WcaOnRails/spec/models/competition_wcif_spec.rb

```diff
@@ -0,0 +1,176 @@
+                "timeLimit" => {
+                  "centiseconds" => 10.minutes.in_centiseconds,
+                  "cumulativeRoundIds" => [],
+                },
+                "cutoff" => nil,
+                "advancementCondition" => nil,
+              },
+            ],
+          },
+        ],
+      )
+    end
+  end
+
+  describe "#set_wcif_events!" do
+    it "removes competition event when wcif rounds are empty" do
+      wcif = competition.to_wcif
+      wcif_444_event = wcif["events"].find { |e| e["id"] == "444" }
+      wcif_444_event["rounds"] = []
+
+      competition.set_wcif_events!(wcif["events"])
+
+      wcif["events"].reject! { |e| e["id"] == "444" }
+      expect(competition.to_wcif["events"]).to eq(wcif["events"])
+      expect(competition.events.map(&:id)).to match_array %w(333)
+    end
+
+    it "removes competition event when wcif event is missing" do
+      wcif = competition.to_wcif
+      wcif["events"].reject! { |e| e["id"] == "444" }
+
+      competition.set_wcif_events!(wcif["events"])
+
+      expect(competition.to_wcif["events"]).to eq(wcif["events"])
+      expect(competition.events.map(&:id)).to match_array %w(333)
+    end
+
+    it "creates competition event when adding round to previously nonexistent event" do
+      wcif = competition.to_wcif
+      wcif["events"] << {
+        "id" => "555",
+        "rounds" => [
+          {
+            "id" => "555-1",
+            "format" => "3",
+            "timeLimit" => {
+              "centiseconds" => 3*60*100,
+              "cumulativeRoundIds" => [],
+            },
+            "cutoff" => nil,
+            "advancementCondition" => nil,
+          },
+        ],
+      }
+
+      competition.set_wcif_events!(wcif["events"])
+
+      expect(competition.to_wcif["events"]).to eq(wcif["events"])
+    end
+
+    it "creates new round when adding round to existing event" do
+      wcif = competition.to_wcif
+      wcif_444_event = wcif["events"].find { |e| e["id"] == "444" }
+      wcif_444_event["rounds"][0]["advancementCondition"] = {
+        "type" => "ranking",
+        "level" => 16,
+      }
+      wcif_444_event["rounds"] << {
+        "id" => "444-2",
+        "format" => "a",
+        "timeLimit" => {
+          "centiseconds" => 10.minutes.in_centiseconds,
+          "cumulativeRoundIds" => [],
+        },
+        "cutoff" => nil,
+        "advancementCondition" => nil,
+      }
+
+      competition.set_wcif_events!(wcif["events"])
+
+      expect(competition.to_wcif["events"]).to eq(wcif["events"])
+
+      # Verify that we can remove the round we just added, so long as we
+      # clear the advancementCondition on the first round.
+      wcif_444_event["rounds"][0]["advancementCondition"] = nil
+      wcif_444_event["rounds"].pop
+      competition.set_wcif_events!(wcif["events"])
+
+      expect(competition.to_wcif["events"]).to eq(wcif["events"])
+    end
+
+    it "can change round format to '3'" do
+      wcif = competition.to_wcif
```

Comment:
I believe you can extract this line from the beginning of each `it` into a `let!` call.